File: essentials/chapter_5/langchain/agent_structured_output.ts
/**
 * Agent Memory Structured Gateway - LangChain TypeScript with structured JSON responses.
 */

import { LangChainLLMManager as Chapter5LangChainManager } from './agent_memory_persist.ts';
import { interactiveCli, parseStructuredJsonResponse, printCliHelp } from '../../../shared/essentials/utils.ts';

export const STRUCTURED_TEMPLATE = `Given the topic below, provide:

1. A direct factual answer (if possible)
2. A summary of what the question is about
3. Relevant keywords
4. A distilled answer (short phrase or value-only form of the answer)

Respond in the following JSON format:
{
  "answer": "...",
  "summary": "...",
  "keywords": ["...", "..."],
  "distilled": "...",
  "metadata": {
    "confidence": "high|medium|low",
    "notes": "optional extra context"
  }
}

Topic: {topic}`;

class LangChainLLMManager extends Chapter5LangChainManager {
    constructor(memoryEnabled = true) {
        super(memoryEnabled);
        this.framework = 'LangChain Structured Output TypeScript';
    }

    _extractObjectFromText(rawResponse, key) {
        const marker = `${key}=`;
        const markerIndex = rawResponse.indexOf(marker);
        if (markerIndex === -1) return null;

        const start = rawResponse.indexOf('{', markerIndex);
        if (start === -1) return null;

        let depth = 0;
        let inString = false;
        let quote = '';
        let escape = false;

        for (let i = start; i < rawResponse.length; i += 1) {
            const ch = rawResponse[i];
            if (inString) {
                if (escape) {
                    escape = false;
                } else if (ch === '\\') {
                    escape = true;
                } else if (ch === quote) {
                    inString = false;
                }
                continue;
            }

            if (ch === '"' || ch === "'") {
                inString = true;
                quote = ch;
                continue;
            }

            if (ch === '{') {
                depth += 1;
            } else if (ch === '}') {
                depth -= 1;
                if (depth === 0) {
                    const block = rawResponse.slice(start, i + 1);
                    try {
                        return JSON.parse(block.replace(/'/g, '"'));
                    } catch {
                        return null;
                    }
                }
            }
        }

        return null;
    }

    _buildMetadata(result, rawResponse) {
        const responseMetadata = result.response_metadata ?? this._extractObjectFromText(rawResponse, 'response_metadata');
        const usageMetadata = result.usage_metadata ?? this._extractObjectFromText(rawResponse, 'usage_metadata');
        const tokenUsageSource = responseMetadata?.token_usage;
        const tokenUsage = tokenUsageSource
            ? {
                completion_tokens: tokenUsageSource.completion_tokens,
                prompt_tokens: tokenUsageSource.prompt_tokens,
                total_tokens: tokenUsageSource.total_tokens,
            }
            : (usageMetadata
                ? {
                    completion_tokens: usageMetadata.output_tokens,
                    prompt_tokens: usageMetadata.input_tokens,
                    total_tokens: usageMetadata.total_tokens,
                }
                : null);

        const metadata = {
            provider: result.provider,
            model: result.model,
            framework: this.framework,
            session_id: result.sessionId,
            temperature: result.temperature,
            max_tokens: result.maxTokens,
            raw_response_chars: rawResponse.length,
            token_usage: tokenUsage ? Object.fromEntries(Object.entries(tokenUsage).filter(([, value]) => value != null)) : null,
        };

        return Object.fromEntries(Object.entries(metadata).filter(([, value]) => value != null));
    }

    async askQuestion(topic, provider = null, template = STRUCTURED_TEMPLATE, maxTokens = 1000, temperature = 0.7, sessionId = 'default') {
        const effectiveTemplate = template === '{topic}' ? STRUCTURED_TEMPLATE : template;
        const result = await super.askQuestion(topic, provider, effectiveTemplate, maxTokens, temperature, sessionId);

        if (!result.success) {
            return result;
        }

        const rawResponse = typeof result.response === 'string' ? result.response : String(result.response ?? '');
        try {
            const parsed = parseStructuredJsonResponse(rawResponse);
            parsed.metadata = {
                ...(parsed.metadata || {}),
                ...this._buildMetadata(result, rawResponse),
            };
            return {
                ...result,
                response: parsed,
                rawAnswer: parsed.answer ?? rawResponse,
            };
        } catch (error) {
            return {
                ...result,
                success: false,
                error: `Failed to parse structured JSON response: ${error.message}`,
                response: null,
                rawResponse,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.includes('-h') || args.includes('--help')) {
        printCliHelp(process.argv[1]);
        return;
    }
    if (args.includes('web')) {
        const { runWebServer } = await import('../../../shared/essentials/web.ts');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}

File: essentials/chapter_5/agent_structured_output.py
"""Agent Memory Structured Gateway - LangChain with structured JSON responses."""

import ast
import os
import sys
from typing import Dict, Optional

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))

sys.path.append(REPO_ROOT)

if any(arg in {"-h", "--help"} for arg in sys.argv[1:]):
    from shared.utils import print_cli_help

    print_cli_help(sys.argv[0])
    sys.exit(0)


from essentials.chapter_5.agent_memory_persist import LangChainLLMManager as Chapter5LangChainManager
from shared.essentials.utils import interactive_cli, parse_structured_json_response


STRUCTURED_TEMPLATE = """
Given the topic below, provide:

1. A direct factual answer (if possible)
2. A summary of what the question is about
3. Relevant keywords
4. A distilled answer (short phrase or value-only form of the answer)

Respond in the following JSON format:
{{
  "answer": "...",
  "summary": "...",
  "keywords": ["...", "..."],
  "distilled": "...",
  "metadata": {{
    "confidence": "high|medium|low",
    "notes": "optional extra context"
  }}
}}

Topic: {topic}
""".strip()


class LangChainLLMManager(Chapter5LangChainManager):
    """Chapter 5 structured manager layered on persistent memory."""

    def __init__(self, memory_enabled: bool = True):
        super().__init__(memory_enabled=memory_enabled)
        self.framework = "LangChain Structured Output"

    @staticmethod
    def _extract_dict_from_text(raw_response: str, key: str) -> Optional[Dict]:
        marker = f"{key}="
        idx = raw_response.find(marker)
        if idx == -1:
            return None

        start = raw_response.find("{", idx)
        if start == -1:
            return None

        depth = 0
        in_string = False
        quote = ""
        escape = False

        for pos in range(start, len(raw_response)):
            ch = raw_response[pos]
            if in_string:
                if escape:
                    escape = False
                elif ch == "\\":
                    escape = True
                elif ch == quote:
                    in_string = False
                continue

            if ch in ('"', "'"):
                in_string = True
                quote = ch
                continue

            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0:
                    try:
                        return ast.literal_eval(raw_response[start : pos + 1])
                    except Exception:
                        return None

        return None

    def _build_metadata(self, result: Dict, raw_response: str) -> Dict:
        response_metadata = result.get("response_metadata")
        usage_metadata = result.get("usage_metadata")

        if response_metadata is None:
            response_metadata = self._extract_dict_from_text(raw_response, "response_metadata")
        if usage_metadata is None:
            usage_metadata = self._extract_dict_from_text(raw_response, "usage_metadata")

        token_usage = None
        if isinstance(response_metadata, dict) and isinstance(response_metadata.get("token_usage"), dict):
            usage = response_metadata.get("token_usage")
            token_usage = {
                "completion_tokens": usage.get("completion_tokens"),
                "prompt_tokens": usage.get("prompt_tokens"),
                "total_tokens": usage.get("total_tokens"),
            }
        elif isinstance(usage_metadata, dict):
            token_usage = {
                "completion_tokens": usage_metadata.get("output_tokens"),
                "prompt_tokens": usage_metadata.get("input_tokens"),
                "total_tokens": usage_metadata.get("total_tokens"),
            }

        if isinstance(token_usage, dict):
            token_usage = {k: v for k, v in token_usage.items() if v is not None}
            if not token_usage:
                token_usage = None

        metadata = {
            "provider": result.get("provider"),
            "model": result.get("model"),
            "framework": self.framework,
            "session_id": result.get("session_id"),
            "temperature": result.get("temperature"),
            "max_tokens": result.get("max_tokens"),
            "raw_response_chars": len(raw_response),
            "token_usage": token_usage,
        }
        return {k: v for k, v in metadata.items() if v is not None}

    def ask_question(
        self,
        topic: str,
        provider: Optional[str] = None,
        template: str = STRUCTURED_TEMPLATE,
        max_tokens: int = 1000,
        temperature: float = 0.7,
        session_id: str = "default",
    ) -> Dict:
        effective_template = STRUCTURED_TEMPLATE if template == "{topic}" else template
        result = super().ask_question(topic, provider, effective_template, max_tokens, temperature, session_id)

        if not result.get("success"):
            return result

        original_response = result.get("response")
        raw_response = original_response if isinstance(original_response, str) else str(original_response)
        metadata_source = original_response if isinstance(original_response, str) else str(original_response)
        try:
            parsed = parse_structured_json_response(raw_response)
            parsed["metadata"] = {
                **(parsed.get("metadata") or {}),
                **self._build_metadata(result, metadata_source),
            }
        except Exception as exc:
            return {
                **result,
                "success": False,
                "error": f"Failed to parse structured JSON response: {exc}",
                "response": None,
                "raw_response": raw_response,
            }

        return {
            **result,
            "response": parsed,
            "raw_answer": parsed.get("answer", raw_response),
        }


def main():
    args = sys.argv[1:]
    if "web" in args:
        from shared.essentials.web import run_web_server

        run_web_server(lambda: LangChainLLMManager(memory_enabled=True))
    else:
        interactive_cli(LangChainLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()
